python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1121,344 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
+        F: typing.Union[dolfinx.fem.Form, ufl.form.Form],
+        u: dolfinx.fem.Function,
+        J: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        P: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        bcs: typing.Optional[list[dolfinx.fem.DirichletBC]] = None,
+        form_compiler_options: typing.Optional[dict] = None,
+        jit_options: typing.Optional[dict] = None,
+    ):
+        """Class for constructing the residual and Jacobian constructors for a SNES problem.
+
+        Args:
+            F: The PDE residual F(u, v)
+            u: The unknown
+            bcs: List of Dirichlet boundary conditions
+            J: UFL representation of the Jacobian (Optional)
+            P: UFL representation of a preconditioner (Optional)
+            form_compiler_options: Options used in FFCx
+                compilation of this form. Run ``ffcx --help`` at the
+                command line to see all available options.
+            jit_options: Options used in CFFI JIT compilation of C
+                code generated by FFCx. See ``python/dolfinx/jit.py``
+                for all available options. Takes priority over all other
+                option values.
+        """
+        self._comm = u.function_space.mesh.comm
+        form_compiler_options = {} if form_compiler_options is None else form_compiler_options
+        jit_options = {} if jit_options is None else jit_options
+
+        self._L = _create_form(
+            F, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+        if J is None:
+            V = u.function_space
+            du = ufl.TrialFunction(V)
+            J = ufl.derivative(F, u, du)
+        self._a_prec = (
+            None
+            if P is None
+            else _create_form(
+                P, form_compiler_options=form_compiler_options, jit_options=jit_options
+            )
+        )
+        self._a = _create_form(
+            J, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+        self.bcs = bcs
+        self.u = u
+
+    @property
+    def comm(self):
+        """The MPI communicator associated with the problem"""
+        return self._comm
+
+    @property
+    def L(self) -> Form:
+        """Compiled linear form (the residual form)"""
+        return self._L
+
+    @property
+    def a(self) -> Form:
+        """Compiled bilinear form (the Jacobian form)"""
+        return self._a
+
+    @property
+    def P(self) -> Form:
+        """Compiled bilinear form (the preconditioner form)"""
+        return self._a_prec
+
+    def update_solution(self, x: PETSc.Vec):
+        """Update the solution vector in the function used in the residual and Jacobian.
+
+        Args:
+            x: The vector containing the latest solution
+        """
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        x.copy(self.u.x.petsc_vec)
```

Comment:
Why is there a copy made? Should the Function be called `replace`? The two ghost updates look weird. Are they necessary?

Reply:
I've changed `update_solution` to replace solution. Yes, both updates are needed (The solver fails to converge otherwise in the tests).